ros/src/computing/perception/localization/packages/ekf_localizer/src/ekf_localizer.cpp

```diff
@@ -0,0 +1,672 @@
+  C(0, IDX::VX) = 1.0; // for vx
+  C(1, IDX::WZ) = 1.0; // for wz
+
+  /* Set measurement noise covariancs */
+  Eigen::MatrixXd R = Eigen::MatrixXd::Zero(dim_y, dim_y);
+  R(0, 0) = twist_stddev_vx_ * twist_stddev_vx_ * ekf_dt_; // for vx
+  R(1, 1) = twist_stddev_wz_ * twist_stddev_wz_ * ekf_dt_; // for wz
+
+  /* In order to avoid a large change by update, measurement update is performed by dividing at every step. */
+  R *= (ekf_rate_ / twist_rate_);
+
+  ekf_.updateWithDelay(y, C, R, delay_step);
+
+  // debug
+  Eigen::MatrixXd X_result(dim_x_, 1);
+  ekf_.getLatestX(X_result);
+  DEBUG_PRINT_MAT(X_result.transpose());
+  DEBUG_PRINT_MAT((X_result - X_curr).transpose());
+};
+
+/*
+ * mahalanobisGate
+ */
+bool EKFLocalizer::mahalanobisGate(const double &dist_max, const Eigen::MatrixXd &x,
+                                       const Eigen::MatrixXd &obj_x, const Eigen::MatrixXd &cov)
+{
+  Eigen::MatrixXd mahalanobis_squared = (x - obj_x).transpose() * cov.inverse() * (x - obj_x);
+  DEBUG_INFO("measurement update: mahalanobis = %f, gate limit = %f", std::sqrt(mahalanobis_squared(0)), dist_max);
+  if (mahalanobis_squared(0) > dist_max * dist_max)
+  {
+    return false;
+  }
+
+  return true;
+}
+
+/*
+ * publishEstimatedPose
+ */
+void EKFLocalizer::publishEstimatedPose()
+{
+  ros::Time current_time = ros::Time::now();
+  Eigen::MatrixXd X(dim_x_, 1);
+  Eigen::MatrixXd P(dim_x_, dim_x_);
+  ekf_.getLatestX(X);
+  ekf_.getLatestP(P);
+
+  /* publish latest pose */
+  pub_pose_.publish(current_ekf_pose_);
+
+  /* publish latest pose with covariance */
+  geometry_msgs::PoseWithCovarianceStamped pose_cov;
+  pose_cov.header.stamp = current_time;
+  pose_cov.header.frame_id = current_ekf_pose_.header.frame_id;
+  pose_cov.pose.pose = current_ekf_pose_.pose;
+  for (int i = 0; i < 36; ++i)
+  {
+    pose_cov.pose.covariance[i] = 0.0;
+  }
+  pose_cov.pose.covariance[0] = P(IDX::X, IDX::X);      // x, x
+  pose_cov.pose.covariance[1] = P(IDX::X, IDX::Y);      // x, y
+  pose_cov.pose.covariance[5] = P(IDX::X, IDX::YAW);    // x, yaw
+  pose_cov.pose.covariance[6] = P(IDX::Y, IDX::X);      // y, x
+  pose_cov.pose.covariance[7] = P(IDX::Y, IDX::Y);      // y, y
+  pose_cov.pose.covariance[11] = P(IDX::Y, IDX::YAW);   // y, yaw
+  pose_cov.pose.covariance[30] = P(IDX::YAW, IDX::X);   // yaw, x
+  pose_cov.pose.covariance[31] = P(IDX::YAW, IDX::Y);   // yaw, y
+  pose_cov.pose.covariance[35] = P(IDX::YAW, IDX::YAW); // yaw, yaw
+  pub_pose_cov_.publish(pose_cov);
+
+  /* publish latest twist */
+  pub_twist_.publish(current_ekf_twist_);
+
+  /* publish yaw bias */
+  std_msgs::Float64 yawb;
+  yawb.data = X(IDX::YAWB);
+  pub_yaw_bias_.publish(yawb);
+
+  /* debug measured pose */
+  if (current_pose_ptr_ != nullptr)
+  {
+    geometry_msgs::PoseStamped p;
+    p = *current_pose_ptr_;
+    p.header.stamp = current_time;
+    pub_measured_pose_.publish(p);
+  }
+
+  /* debug publish */
+  double RAD2DEG = 180.0 / 3.141592;
```

Comment:
why not use [mathutils](https://github.com/autowarefoundation/autoware/blob/master/ros/src/common/libs/amathutils_lib/include/amathutils_lib/amathutils.hpp#L18) if you already included it?

Reply:
There were some discusssion for amathutils.hpp (#2045). I'm not sure it will be removed or modified, so I didn't want to depend on it.